besu/src/main/java/org/hyperledger/besu/cli/subcommands/storage/TrieLogHelper.java

```diff
@@ -0,0 +1,227 @@
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.google.common.base.Splitter;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.bouncycastle.util.encoders.Base64;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Helper class for counting and pruning trie logs */
+public class TrieLogHelper {
+  private static final String trieLogFile = "trieLogsToRetain.txt";
+  private static final Logger LOG = LoggerFactory.getLogger(TrieLogHelper.class);
+
+  static void prune(
+      final PrintWriter out,
+      final DataStorageConfiguration config,
+      final BonsaiWorldStateKeyValueStorage rootWorldStateStorage,
+      final MutableBlockchain blockchain)
+      throws IOException {
+
+    TrieLogHelper.validatePruneConfiguration(config);
+    final long layersToRetain = config.getUnstable().getBonsaiTrieLogRetentionThreshold();
+    final long chainHeight = blockchain.getChainHeadBlockNumber();
+    final long lastBlockToRetainTrieLogsFor = chainHeight - layersToRetain;
+    final Optional<Hash> finalizedBlockHash = blockchain.getFinalized();
+    if (finalizedBlockHash.isEmpty()) {
+      LOG.error("No finalized block present, skipping pruning");
+      return;
+    } else {
+      if (blockchain.getBlockHeader(finalizedBlockHash.get()).get().getNumber()
+          < lastBlockToRetainTrieLogsFor) {
+        LOG.error("Trying to prune more layers than the finalized block height, skipping pruning");
+        return;
+      }
+    }
+
+    // retrieve the layersToRetains hashes from blockchain
+    final List<Hash> hashesToRetain = new ArrayList<>();
+
+    for (long i = chainHeight; i > lastBlockToRetainTrieLogsFor; i--) {
+      final Optional<BlockHeader> header = blockchain.getBlockHeader(i);
+      header.ifPresent(blockHeader -> hashesToRetain.add(blockHeader.getHash()));
+    }
+
+    IdentityHashMap<byte[], byte[]> trieLogsToRetain;
+    if ((long) hashesToRetain.size() == layersToRetain) {
+      trieLogsToRetain = new IdentityHashMap<>();
+      // save trielogs in a flatfile as a fail-safe
+      out.println("Obtaining trielogs to retain...");
+      hashesToRetain.forEach(
+          hash -> {
+            rootWorldStateStorage
+                .getTrieLog(hash)
+                .ifPresent(trieLog -> trieLogsToRetain.put(hash.toArrayUnsafe(), trieLog));
+          });
+      out.println("Saving trielogs to retain in file...");
+      saveTrieLogsInFile(trieLogsToRetain);
+    } else {
+      // try to read the triLogs from the flatfile
+      trieLogsToRetain = readTrieLogsFromFile();
```

Comment:
Don't understand how reading from the file helps in this case. We didn't find all the hashes but also haven't created the file at this point either unless we are relying running the command again to read the previous state

Reply:
I think this is a failsafe in for a subsequent execution where something went terribly wrong during the truncation and restore-from-file process.

what would be nice, but not idiomatic, is if we had a few "are you sure?" or "do you want to?" prompts that could be bypassed by `-y` similar to ubuntu apt commands